### PR TITLE
add docker support

### PR DESCRIPTION
**Pull request by chenshuai2144**
@ja573  made some changes
move `Dockerfile` and `Dockerfile.dev` to `/`
move test:all to package.json

### CODE REVIEW

**Comment by chenshuai2144** [comment]
I should roll back this

**Comment by ja573** [comment]
> I should roll back this

Until we find out what's causing https://github.com/umijs/umi/issues/1161 the flag is actually very helpful

**Comment by ja573** [comment]
In fact the flag hack seems to fix the memory leak when testing in circleci and appveyor - so it might be better to leave it here for the time being. Once the problem is fixed in umi it can be removed

**Comment by chenshuai2144** [comment]
mi keeps refreshing the shell, this is his feature, judging that this is the hack code itself.

**Comment by ja573** [comment]
> mi keeps refreshing the shell, this is his feature.
> Use shell output to determine if the service startup is successful. This method is very hacky.

I'm not entirely sure I understand what you mean @chenshuai2144 , sorry - could you rephrase the comment a bit?

**Comment by chenshuai2144** [comment]
I think we should judge whether the service is started by other means.

File: docker/nginx.conf [hunk 1]
```diff
@@ -1,5 +1,12 @@
 server {
     listen 80;
+    # gzip config
+    gzip on;
+    gzip_min_length 1k;
+    gzip_comp_level 9
```
Comment (ja573) [inline docker/nginx.conf]: this statement is missing a semicolon

**Comment by ja573** [comment]
> I think we should judge whether the service is started by other means.

That's fair enough

File: Dockerfile [hunk 2]
```diff
@@ -7,7 +7,5 @@ RUN npm install --silent --no-cache
 
 COPY ./ ./
 
-RUN sh ./tests/fix_puppeteer.sh
```
Comment (ja573) [inline Dockerfile]: I'm sure I commented about this somewhere, but I cannot find it. Running tests upon build is quite useful, as it will break the build if they do not succeed

File: package.json [hunk 3]
```diff
@@ -17,10 +17,13 @@
...
     "test": "umi test",
-    "generate-mock": "node ./scripts/generateMock.js",
     "test:component": "umi test ./src/components",
     "test:all": "node ./tests/run-tests.js",
-    "prettier": "prettier --write ./src/**/**/**/*"
+    "prettier": "prettier --write ./src/**/**/**/*",
+    "docker:dev":"docker-compose -f ./docker/docker-compose.dev.yml up",
+    "docker:build":"docker-compose -f ./docker/docker-compose.dev.yml build",
+    "docker-prod:dev":"docker-compose -f ./docker/docker-compose.yml up",
+    "docker-prod:build":"npm run test:all && docker-compose -f ./docker/docker-compose.yml build"
```
Comment (ja573) [inline package.json]: dito

**Comment by chenshuai2144** [comment]
i will use a container with puppeteer
